tpu-client-next/src/workers_cache.rs

```diff
@@ -86,99 +80,113 @@ pub enum WorkersCacheError {
-
-            if let Err(WorkersCacheError::ReceiverDropped) = send_res {
-                // Remove the worker from the cache, if the peer has disconnected.
-                if let Some(current_worker) = workers.pop(peer) {
-                    // To avoid obscuring the error from send, ignore a possible
-                    // `TaskJoinFailure`.
-                    let close_result = current_worker.shutdown().await;
-                    if let Err(error) = close_result {
-                        error!("Error while closing worker: {error}.");
-                    }
-                }
-            }
+        );
+        let send_res = current_worker.try_send_transactions(txs_batch);
 
-            send_res
-        };
-
-        tokio::select! {
-            send_res = body => send_res,
-            () = cancel.cancelled() => Err(WorkersCacheError::ShutdownError),
+        if let Err(WorkersCacheError::ReceiverDropped) = send_res {
+            warn!(
+                "Failed to deliver transaction batch for leader {}, drop batch.",
+                peer.ip()
+            );
         }
-    }
 
-    pub fn transaction_stats(&self) -> &HashMap<IpAddr, SendTransactionStats> {
-        &self.send_stats_per_addr
+        send_res
     }
 
     /// Closes and removes all workers in the cache. This is typically done when
     /// shutting down the system.
-    pub async fn shutdown(&mut self) {
-        // Interrupt any outstanding `send_txs()` calls.
+    pub(crate) async fn shutdown(&mut self) {
+        // Interrupt any outstanding `send_transactions()` calls.
         self.cancel.cancel();
 
         while let Some((leader, worker)) = self.workers.pop_lru() {
-            self.shutdown_worker(leader, worker).await;
+            let res = worker.shutdown().await;
+            if let Err(err) = res {
+                debug!("Error while shutting down worker for {leader}: {err}");
+            }
         }
     }
+}
 
-    /// Shuts down a worker for a given peer by closing the worker and gathering
-    /// its transaction statistics.
-    async fn shutdown_worker(&mut self, leader: SocketAddr, worker: WorkerInfo) {
-        let res = worker.shutdown().await;
+/// [`ShutdownWorker`] takes care of stopping the worker. It's method
+/// `shutdown()` should be executed in a separate task to hide the latency of
+/// finishing worker gracefully.
+pub(crate) struct ShutdownWorker {
+    leader: SocketAddr,
+    worker: WorkerInfo,
+}
 
-        let stats = match res {
-            Ok(stats) => stats,
-            Err(err) => {
+impl ShutdownWorker {
+    pub(crate) fn leader(&self) -> SocketAddr {
+        self.leader
+    }
+
+    pub(crate) async fn shutdown(self) -> Result<(), WorkersCacheError> {
+        self.worker.shutdown().await
+    }
+}
+
+pub(crate) fn maybe_shutdown_worker(worker: Option<ShutdownWorker>) {
+    if let Some(worker) = worker {
+        tokio::spawn(async move {
```

Comment:
Spawning tasks without observing their execution, I think, is a bit fragile.

The executor will keep track of them, but it is not clear in the rest of the code as to what is going on.
And if those tasks take a long time to complete or even hang, it would be just reflected in the executor taking a long time to finish or hanging.

I do not fully understand the end to end logic here.
So, it could be hard to track.
But I suggest you consider putting all the shutdown tasks in an `UnorderedFutures` or something like that, allowing you to track and even interrupt the shutdown process.

In particular, it would allow the shutdown errors to be processed in a centralized location, rather than them being just printed as warnings.
Though again, not sure if you really need this feature.
But, maybe for collection of the stats?

Reply:
> But I suggest you consider putting all the shutdown tasks in an
> UnorderedFutures or something like that, allowing you to track and even
> interrupt the shutdown process.

we explicitly don't want to do this tho: we don't want cleanup to delay sending
transactions. So we do need tasks. We could have a JoinSet and call
poll_join_next to poll without blocking and pop tasks off the join set, although
that seems extra work for little gain. Background tasks don't block the runtime https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#shutdown